### ISSUE TITLE
super slow when zoomed in with disappearing elements

### ISSUE THREAD

**Issue by Mrazator** [excalidraw/excalidraw#8242]
recording while zooming in with trackpad:
https://github.com/user-attachments/assets/7cc12287-fb76-4d9e-bafd-f886ee64522d

repro scene:
https://excalidraw.com/#json=1SbjsdotedadDg7pM4Ahp,z9uJCfl5W8d73Id6x_xVcw

system info:
- Arc 1.51.0
- MacOS: 14.2.1

**Comment by BlueGreenMagick** [comment]
From what I can tell, the problem is  that rendering huge arrows with text is especially expensive in Firefox (Mac M1). It seems fine on Chrome and Safari.

**Comment by dwelle** [comment]
That's right, it's due to the labeled arrow there. Clipping is slow. But apparently, it's completely killing performance if the clip area is large enough, such as when you zoom the labeled arrow.

I don't have any ideas here.

@ad1992 was suggesting we don't use clipping at all, and instead split the arrow into individual parts, but IMO it's not feasible as it gets insanely complex fast. At best we could do this only for straight, 2-point arrows, and even then we couldn't retain rendering parity due to how rough.js renders different-length lines:

![image](https://github.com/user-attachments/assets/cf03f2c8-6dad-4a27-b687-da3f2e993a07)

**Comment by BlueGreenMagick** [comment]
I could think of 2 possible solutions to improve the performance.

Problematic code:

https://github.com/excalidraw/excalidraw/blob/df8875a497b2111ec8c21963d66a2e2396d15bbd/packages/excalidraw/renderer/renderElement.ts#L483-L485


#### 1. tempCanvas as viewport

Instead of drawing the whole scaled arrow to `tempCanvas`, we can draw only the 'visible' part of arrow to `tempCanvas`. So `tempCanvas` would have identical dimension as static canvas dimension.

Working example: 
https://github.com/BlueGreenMagick/excalidraw/blob/6c2a8fd3ef0babbb7a786fb1b0f00ffce7537025/packages/excalidraw/renderer/renderElement.ts#L516

For a large (zoomed in) arrow, it decreases `drawElementFromCanvas()` cost from 100ms+ to ~1ms.

However, the performance is significantly worsened for small labelled arrows. So this optimization should only be applied if arrow is larger than root canvas.

#### 2. Cache tempCanvas

Simply caching `tempCache` that holds labeled arrow would improve performance drastically for moving around.
